### PR TITLE
Support non-QWERTY keyboards

### LINKED ISSUE #4144: Text input is hard-wired to QWERTY
It seems that Servo receives keyboard event based on physical keys, and has its own interpretation of how to map they to text which assumes a QWERTY keyboard.

The fix is **not** to add support for more keyboard layouts. Servo should not be in the business of maintaining a repository of all keyboard layouts in the world, this is the job of the operating system. Servo should have the operating system give it Unicode text in a way that supports whatever [input method (a.k.a. IME)](https://en.wikipedia.org/wiki/Input_method) the operating system may or may not have.

CC @jdm

### LINKED ISSUE #11991: Keyboard keys AZERTY are badly mapped
When i am on the start page, i want to put and URL, but I cannot, as keys are not correctly mapped.

OS : Kubuntu 14.04 LTS , arch 64 . Keyboard layout : AZERTY, 105 keys

Letters seems corrects, but symbols keys are bad : 

numeric rank:
«²», «&» , «é», «"», «(», «è», «_», «ç», «à», «)» are not mapped

right side of the two upper letter ranks : none of them

On lower rank, after VBN letters : 
hitting «?» gives a «<» , non numpad «.» gives a «:» , «/» and «§» are mapped to nothing,

### PR DESCRIPTION
**Pull request by jdm**
Using the ReceivedCharacter event from glutin, we can obtain the actual key characters that the user is pressing and releasing. This gets passed to the script thread along with the physical key data, since KeyboardEvent needs both pieces of information, where they get merged into a single logical key that gets processed by clients like TextInput without any special changes.

Tested by switching my macbook keyboard to dvorak and looking at the output of keypress/keyup/keydown event listeners, as well as playing with tests/html/textarea.html. Non-content keybindings like reload work as expected, too - the remapped keybinding triggers the reload action.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #4144 
- [X] These changes do not require tests because I can't think of a way to test remapped keyboard input

Fixes  #11991.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11950)

### CODE REVIEW

**Comment by nox** [comment]
Mac French keyboard here, 'a' works, but not shift etc to input '.' with 'shift-;'.

**Comment by paulrouget** [comment]
> Mac French keyboard here, 'a' works, but not shift etc to input '.' with 'shift-;'.

What happens if you press "Shift + a"?

**Comment by jdm** [comment]
Tested against a french keyboard and everything seems to be working as expected.

File: ports/glutin/window.rs [hunk 1]
```diff
@@ -233,7 +241,13 @@ impl Window {
 
     fn handle_window_event(&self, event: glutin::Event) -> bool {
         match event {
-            Event::KeyboardInput(element_state, _scan_code, Some(virtual_key_code)) => {
+            Event::ReceivedCharacter(ch) => {
+                assert!(self.pending_key_event_char.get().is_none());
```
Comment (mbrubeck) [inline ports/glutin/window.rs]: This assertion panics if I try to type a character using the Compose key (e.g. Compose+A+E to type `æ`).

**Comment by mbrubeck** [comment]
My Linux/X11 dvorak key mapping still works with this patch, except for one corner case noted above.

**Comment by nox** [comment]
@jdm It seems you didn't include my bincode bump that fixes the char serialization.

**Comment by jdm** [comment]
@nox It's... right there? In the PR?

**Comment by nox** [comment]
Oh right, didn't look at the right commit.

File: components/script/dom/keyboardevent.rs [hunk 2]
```diff
@@ -69,6 +70,7 @@ impl KeyboardEvent {
                cancelable: bool,
                view: Option<&Window>,
                _detail: i32,
+               ch: Option<char>,
```
Comment (nox) [inline components/script/dom/keyboardevent.rs]: Why do we need both this and `char_code`?

Comment (jdm) [inline components/script/dom/keyboardevent.rs]: Because char_code isn't present in all key events, per spec.

**Comment by pcwalton** [comment]
@jdm Who would you like to review this?

**Comment by jdm** [comment]
Someone who wants to learn more about how we do keyboard input? I don't have a name in mind, unfortunately.

**Comment by nox** [comment]
SGTM. What do we need for this to get landed? What's blocking this to be reviewed?

**Comment by jdm** [comment]
Nothing's blocking it that I'm aware.

**Comment by emilio** [comment]
Looks good to me modulo the nits, the ipc-channel/bincode update, and the compose key, that is https://github.com/servo/glutin/pull/101

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 12 of 12 files at r1.
Review status: 4 of 17 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[components/constellation/constellation.rs, line 809 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11950#-KLsEXIa-UvJOPK2GV25:-KLsEXIa-UvJOPK2GV26:-1682507779) ([raw file](https://github.com/servo/servo/blob/373e3d5f837b57e03a6ce211716c2ed8089f5eca/components/constellation/constellation.rs#L809)):*
> ```Rust
>             FromScriptMsg::SendKeyEvent(key, key_state, key_modifiers, ch) => {
>                 // FIXME: https://github.com/servo/ipc-channel/issues/84
>                 let ch = ch.map(|ch| ch as char);
> ```

This can be removed now right?

---

*[components/constellation/constellation.rs, line 1636 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11950#-KLsEbY_R1SQ_n0n_5CC:-KLsEbY_R1SQ_n0n_5CD:-1713162174) ([raw file](https://github.com/servo/servo/blob/373e3d5f837b57e03a6ce211716c2ed8089f5eca/components/constellation/constellation.rs#L1636)):*
> ```Rust
>                 };
>                 for (key, mods, state) in cmd {
>                     let event = CompositorEvent::KeyEvent(key, state, mods, None);
> ```

Can we be consistent with the order here?

Either `(ch, key, state, mods)` or `(key, state, mods, ch)`.

---

*[ports/glutin/window.rs, line 245 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11950#-KLVRgYY1POAwrEbDbLg-r3-245:-KLsF2A-0JOdkoIp8ts9:498935693) ([raw file](https://github.com/servo/servo/blob/d5a0204e815373a006d05da9964667b8add3fc33/ports/glutin/window.rs#L245)):*
<details><summary><i>Previously, mbrubeck (Matt Brubeck) wrote…</i></summary>
> This assertion panics if I try to type a character using the Compose key (e.g. Compose+A+E to type `æ`).
</details>
As Matt, I also happen to hit this assertion whenever I type a spanish character with the compose key (e.g. Compose + ' + a -> `á`).

It seems that glutin doesn't handle the specific compose key. I have a patch to add `VirtualKeyCode::Compose` to glutin, and then we should handle that in the key modifiers ignoring any character that arrives while pressing `Composed` (the actual character will arrive later).

Glutin also sends the composed character twice. I can look deeper into it, but meanwhile we can just check for `pending_key_event_char.get() != Some(ch)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11950#-:-KLsGtBooFr04-kyr-jK:764992869)*

**Comment by emilio** [comment]
Ah, and I don't think I need to say this, but awesome work here, thanks for doing this Josh! :)

**Comment by jdm** [comment]
Looks like you didn't look at the second commit :)

**Comment by emilio** [comment]
Whoops, true, only the compose thing and the order of fields then :)

---

Reviewed 8 of 8 files at r2.
Review status: 12 of 17 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11950#-:-KLsKhEvqAadjYpFaxF0:-298101642)*

**Comment by emilio** [comment]
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11950)*

**Comment by jdm** [comment]
I don't have a machine that will let me test any implementation that deals with the compose key. I'd rather leave that work to someone that can, if that's alright. I went ahead and removed the invalid assertion.

**Comment by emilio** [comment]
Fair enough, I'll do it after this lands. If that's the only thing that has changed (plus addressing review comments), then feel free to r=me.

Otherwise I'll re-review this later.

**Comment by jdm** [comment]
@bors-servo: r=emilio
Thanks!

**Comment by emilio** [comment]
Hm... need to update the cef port :/

**Comment by KiChjang** [comment]
@bors-servo retry
* #11631

**Pull request merged**